// @effect/stream/mjs/internal/take.mjs
var _a;
import * as Chunk from "@effect/data/Chunk";
import { constFalse, constTrue, dual } from "@effect/data/Function";
import * as Option from "@effect/data/Option";
import { pipeArguments } from "@effect/data/Pipeable";
import * as Cause from "@effect/io/Cause";
import * as Effect from "@effect/io/Effect";
import * as Exit from "@effect/io/Exit";
/** @internal */
const TakeSymbolKey = "@effect/stream/Take";
/** @internal */
export const TakeTypeId = /*#__PURE__*/Symbol.for(TakeSymbolKey);
/** @internal */
const takeVariance = {
  _E: _ => _,
  _A: _ => _
};
/** @internal */
export class TakeImpl {
  constructor(exit) {
    this.exit = exit;
    this[_a] = takeVariance;
  }
  pipe() {
    return pipeArguments(this, arguments);
  }
}
_a = TakeTypeId;
/** @internal */
export const chunk = chunk => new TakeImpl(Exit.succeed(chunk));
/** @internal */
export const die = defect => new TakeImpl(Exit.die(defect));
/** @internal */
export const dieMessage = message => new TakeImpl(Exit.die(Cause.RuntimeException(message)));
/** @internal */
export const done = self => Effect.suspend(() => self.exit);
/** @internal */
export const end = /*#__PURE__*/new TakeImpl( /*#__PURE__*/Exit.fail( /*#__PURE__*/Option.none()));
/** @internal */
export const fail = error => new TakeImpl(Exit.fail(Option.some(error)));
/** @internal */
export const failCause = cause => new TakeImpl(Exit.failCause(Cause.map(Option.some)(cause)));
/** @internal */
export const fromEffect = effect => Effect.matchCause(effect, {
  onFailure: failCause,
  onSuccess: of
});
/** @internal */
export const fromExit = exit => new TakeImpl(Exit.mapBoth({
  onFailure: Option.some,
  onSuccess: Chunk.of
})(exit));
/** @internal */
export const fromPull = pull => Effect.matchCause(pull, {
  onFailure: cause => Option.match(Cause.flipCauseOption(cause), {
    onNone: () => end,
    onSome: failCause
  }),
  onSuccess: chunk
});
/** @internal */
export const isDone = self => Exit.match(self.exit, {
  onFailure: cause => Option.isNone(Cause.flipCauseOption(cause)),
  onSuccess: constFalse
});
/** @internal */
export const isFailure = self => Exit.match(self.exit, {
  onFailure: cause => Option.isSome(Cause.flipCauseOption(cause)),
  onSuccess: constFalse
});
/** @internal */
export const isSuccess = self => Exit.match(self.exit, {
  onFailure: constFalse,
  onSuccess: constTrue
});
/** @internal */
export const make = exit => new TakeImpl(exit);
/** @internal */
export const match = /*#__PURE__*/dual(2, (self, {
  onEnd,
  onFailure,
  onSuccess
}) => Exit.match(self.exit, {
  onFailure: cause => Option.match(Cause.flipCauseOption(cause), {
    onNone: onEnd,
    onSome: onFailure
  }),
  onSuccess
}));
/** @internal */
export const matchEffect = /*#__PURE__*/dual(2, (self, {
  onEnd,
  onFailure,
  onSuccess
}) => Exit.matchEffect(self.exit, {
  onFailure: cause => Option.match(Cause.flipCauseOption(cause), {
    onNone: onEnd,
    onSome: onFailure
  }),
  onSuccess
}));
/** @internal */
export const map = /*#__PURE__*/dual(2, (self, f) => new TakeImpl(Exit.map(Chunk.map(f))(self.exit)));
/** @internal */
export const of = value => new TakeImpl(Exit.succeed(Chunk.of(value)));
/** @internal */
export const tap = /*#__PURE__*/dual(2, (self, f) => Effect.asUnit(Exit.forEachEffect(f)(self.exit)));
