// @effect/stm/mjs/internal/tQueue.mjs
var _a, _b;
import * as Chunk from "@effect/data/Chunk";
import { dual } from "@effect/data/Function";
import * as Option from "@effect/data/Option";
import * as RA from "@effect/data/ReadonlyArray";
import * as core from "@effect/stm/internal/core";
import * as OpCodes from "@effect/stm/internal/opCodes/strategy";
import * as stm from "@effect/stm/internal/stm";
import * as tRef from "@effect/stm/internal/tRef";
import * as STM from "@effect/stm/STM";
const TEnqueueSymbolKey = "@effect/stm/TQueue/TEnqueue";
/** @internal */
export const TEnqueueTypeId = /*#__PURE__*/Symbol.for(TEnqueueSymbolKey);
const TDequeueSymbolKey = "@effect/stm/TQueue/TDequeue";
/** @internal */
export const TDequeueTypeId = /*#__PURE__*/Symbol.for(TDequeueSymbolKey);
/** @internal */
export const BackPressure = {
  _tag: OpCodes.OP_BACKPRESSURE_STRATEGY
};
/** @internal */
export const Dropping = {
  _tag: OpCodes.OP_DROPPING_STRATEGY
};
/** @internal */
export const Sliding = {
  _tag: OpCodes.OP_SLIDING_STRATEGY
};
/** @internal */
export const tDequeueVariance = {
  _Out: _ => _
};
/** @internal */
export const tEnqueueVariance = {
  _In: _ => _
};
class TQueueImpl {
  constructor(ref, requestedCapacity, strategy) {
    this.ref = ref;
    this.requestedCapacity = requestedCapacity;
    this.strategy = strategy;
    this[_a] = tDequeueVariance;
    this[_b] = tEnqueueVariance;
    this.size = core.withSTMRuntime(runtime => {
      const queue = tRef.unsafeGet(this.ref, runtime.journal);
      if (queue === undefined) {
        return STM.interruptAs(runtime.fiberId);
      }
      return core.succeed(queue.length);
    });
    this.isFull = core.map(this.size, size => size === this.requestedCapacity);
    this.isEmpty = core.map(this.size, size => size === 0);
    this.shutdown = core.withSTMRuntime(runtime => {
      tRef.unsafeSet(this.ref, void 0, runtime.journal);
      return stm.unit;
    });
    this.isShutdown = core.effect(journal => {
      const queue = tRef.unsafeGet(this.ref, journal);
      return queue === undefined;
    });
    this.awaitShutdown = core.flatMap(this.isShutdown, isShutdown => isShutdown ? stm.unit : core.retry);
    this.peek = core.withSTMRuntime(runtime => {
      const queue = tRef.unsafeGet(this.ref, runtime.journal);
      if (queue === undefined) {
        return core.interruptAs(runtime.fiberId);
      }
      const head = queue[0];
      if (head === undefined) {
        return core.retry;
      }
      return core.succeed(head);
    });
    this.peekOption = core.withSTMRuntime(runtime => {
      const queue = tRef.unsafeGet(this.ref, runtime.journal);
      if (queue === undefined) {
        return core.interruptAs(runtime.fiberId);
      }
      return core.succeed(Option.fromNullable(queue[0]));
    });
    this.take = core.withSTMRuntime(runtime => {
      const queue = tRef.unsafeGet(this.ref, runtime.journal);
      if (queue === undefined) {
        return core.interruptAs(runtime.fiberId);
      }
      const dequeued = queue.shift();
      if (dequeued === undefined) {
        return core.retry;
      }
      tRef.unsafeSet(this.ref, queue, runtime.journal);
      return core.succeed(dequeued);
    });
    this.takeAll = core.withSTMRuntime(runtime => {
      const queue = tRef.unsafeGet(this.ref, runtime.journal);
      if (queue === undefined) {
        return core.interruptAs(runtime.fiberId);
      }
      tRef.unsafeSet(this.ref, [], runtime.journal);
      return core.succeed(queue);
    });
  }
  capacity() {
    return this.requestedCapacity;
  }
  offer(value) {
    return core.withSTMRuntime(runtime => {
      const queue = tRef.unsafeGet(runtime.journal)(this.ref);
      if (queue === undefined) {
        return core.interruptAs(runtime.fiberId);
      }
      if (queue.length < this.requestedCapacity) {
        queue.push(value);
        tRef.unsafeSet(this.ref, queue, runtime.journal);
        return core.succeed(true);
      }
      switch (this.strategy._tag) {
        case OpCodes.OP_BACKPRESSURE_STRATEGY:
          {
            return core.retry;
          }
        case OpCodes.OP_DROPPING_STRATEGY:
          {
            return core.succeed(false);
          }
        case OpCodes.OP_SLIDING_STRATEGY:
          {
            const dequeued = queue.shift();
            if (dequeued === undefined) {
              return core.succeed(true);
            }
            queue.push(value);
            tRef.unsafeSet(this.ref, queue, runtime.journal);
            return core.succeed(true);
          }
      }
    });
  }
  offerAll(iterable) {
    return core.withSTMRuntime(runtime => {
      const as = Array.from(iterable);
      const queue = tRef.unsafeGet(this.ref, runtime.journal);
      if (queue === undefined) {
        return core.interruptAs(runtime.fiberId);
      }
      if (queue.length + as.length <= this.requestedCapacity) {
        tRef.unsafeSet(this.ref, [...queue, ...as], runtime.journal);
        return core.succeed(true);
      }
      switch (this.strategy._tag) {
        case OpCodes.OP_BACKPRESSURE_STRATEGY:
          {
            return core.retry;
          }
        case OpCodes.OP_DROPPING_STRATEGY:
          {
            const forQueue = as.slice(0, this.requestedCapacity - queue.length);
            tRef.unsafeSet(this.ref, [...queue, ...forQueue], runtime.journal);
            return core.succeed(false);
          }
        case OpCodes.OP_SLIDING_STRATEGY:
          {
            const forQueue = as.slice(0, this.requestedCapacity - queue.length);
            const toDrop = queue.length + forQueue.length - this.requestedCapacity;
            const newQueue = queue.slice(toDrop);
            tRef.unsafeSet(this.ref, [...newQueue, ...forQueue], runtime.journal);
            return core.succeed(true);
          }
      }
    });
  }
  takeUpTo(max) {
    return core.withSTMRuntime(runtime => {
      const queue = tRef.unsafeGet(this.ref, runtime.journal);
      if (queue === undefined) {
        return core.interruptAs(runtime.fiberId);
      }
      const [toTake, remaining] = Chunk.splitAt(Chunk.unsafeFromArray(queue), max);
      tRef.unsafeSet(this.ref, Array.from(remaining), runtime.journal);
      return core.succeed(Array.from(toTake));
    });
  }
}
_a = TDequeueTypeId, _b = TEnqueueTypeId;
/** @internal */
export const isTQueue = u => {
  return isTEnqueue(u) && isTDequeue(u);
};
/** @internal */
export const isTEnqueue = u => {
  return typeof u === "object" && u != null && TEnqueueTypeId in u;
};
/** @internal */
export const isTDequeue = u => {
  return typeof u === "object" && u != null && TDequeueTypeId in u;
};
/** @internal */
export const awaitShutdown = self => self.awaitShutdown;
/** @internal */
export const bounded = requestedCapacity => makeQueue(requestedCapacity, BackPressure);
/** @internal */
export const capacity = self => {
  return self.capacity();
};
/** @internal */
export const dropping = requestedCapacity => makeQueue(requestedCapacity, Dropping);
/** @internal */
export const isEmpty = self => self.isEmpty;
/** @internal */
export const isFull = self => self.isFull;
/** @internal */
export const isShutdown = self => self.isShutdown;
/** @internal */
export const offer = /*#__PURE__*/dual(2, (self, value) => self.offer(value));
/** @internal */
export const offerAll = /*#__PURE__*/dual(2, (self, iterable) => self.offerAll(iterable));
/** @internal */
export const peek = self => self.peek;
/** @internal */
export const peekOption = self => self.peekOption;
/** @internal */
export const poll = self => core.map(RA.head)(self.takeUpTo(1));
/** @internal */
export const seek = /*#__PURE__*/dual(2, (self, predicate) => seekLoop(self, predicate));
const seekLoop = (self, predicate) => core.flatMap(self.take, a => predicate(a) ? core.succeed(a) : seekLoop(self, predicate));
/** @internal */
export const shutdown = self => self.shutdown;
/** @internal */
export const size = self => self.size;
/** @internal */
export const sliding = requestedCapacity => makeQueue(requestedCapacity, Sliding);
/** @internal */
export const take = self => self.take;
/** @internal */
export const takeAll = self => self.takeAll;
/** @internal */
export const takeBetween = /*#__PURE__*/dual(3, (self, min, max) => stm.suspend(() => {
  const takeRemainder = (min, max, acc) => {
    if (max < min) {
      return core.succeed(acc);
    }
    return core.flatMap(taken => {
      const remaining = min - taken.length;
      if (remaining === 1) {
        return core.map(a => Chunk.append(a)(Chunk.appendAll(Chunk.unsafeFromArray(taken))(acc)))(self.take);
      }
      if (remaining > 1) {
        return core.flatMap(a => takeRemainder(remaining - 1, max - taken.length - 1, Chunk.append(a)(Chunk.appendAll(Chunk.unsafeFromArray(taken))(acc))))(self.take);
      }
      return core.succeed(Chunk.appendAll(Chunk.unsafeFromArray(taken))(acc));
    })(self.takeUpTo(max));
  };
  return core.map(takeRemainder(min, max, Chunk.empty()), c => Array.from(c));
}));
/** @internal */
export const takeN = /*#__PURE__*/dual(2, (self, n) => takeBetween(n, n)(self));
/** @internal */
export const takeUpTo = /*#__PURE__*/dual(2, (self, max) => self.takeUpTo(max));
/** @internal */
export const unbounded = () => makeQueue(Number.MAX_SAFE_INTEGER, Dropping);
const makeQueue = (requestedCapacity, strategy) => core.map(tRef.make([]), ref => new TQueueImpl(ref, requestedCapacity, strategy));
