// @effect/stm/mjs/internal/tRandom.mjs
var _a;
import * as Context from "@effect/data/Context";
import * as Random from "@effect/data/DeterministicRandom";
import * as Layer from "@effect/io/Layer";
import * as core from "@effect/stm/internal/core";
import * as stm from "@effect/stm/internal/stm";
import * as tArray from "@effect/stm/internal/tArray";
import * as tRef from "@effect/stm/internal/tRef";
const TRandomSymbolKey = "@effect/stm/TRandom";
/** @internal */
export const TRandomTypeId = /*#__PURE__*/Symbol.for(TRandomSymbolKey);
const randomInteger = state => {
  const prng = new Random.PCGRandom();
  prng.setState(state);
  return [prng.integer(0), prng.getState()];
};
const randomIntegerBetween = (low, high) => {
  return state => {
    const prng = new Random.PCGRandom();
    prng.setState(state);
    return [prng.integer(high - low) + low, prng.getState()];
  };
};
const randomNumber = state => {
  const prng = new Random.PCGRandom();
  prng.setState(state);
  return [prng.number(), prng.getState()];
};
const withState = (state, f) => {
  return tRef.modify(f)(state);
};
const shuffleWith = (iterable, nextIntBounded) => {
  const swap = (buffer, index1, index2) => core.flatMap(tmp => core.zipRight(tArray.update(index2, () => tmp)(buffer))(tArray.updateSTM(index1, () => tArray.get(index2)(buffer))(buffer)))(tArray.get(index1)(buffer));
  return core.flatMap(buffer => {
    const array = [];
    for (let i = array.length; i >= 2; i = i - 1) {
      array.push(i);
    }
    return core.zipRight(tArray.toArray(buffer))(stm.forEach(n => core.flatMap(k => swap(buffer, n - 1, k))(nextIntBounded(n)), {
      discard: true
    })(array));
  })(tArray.fromIterable(iterable));
};
/** @internal */
export const Tag = /*#__PURE__*/Context.Tag();
class TRandomImpl {
  constructor(state) {
    this.state = state;
    this[_a] = TRandomTypeId;
    this.next = withState(this.state, randomNumber);
    this.nextBoolean = core.flatMap(this.next, n => core.succeed(n > 0.5));
    this.nextInt = withState(this.state, randomInteger);
  }
  nextRange(min, max) {
    return core.flatMap(this.next, n => core.succeed((max - min) * n + min));
  }
  nextIntBetween(low, high) {
    return withState(this.state, randomIntegerBetween(low, high));
  }
  shuffle(elements) {
    return shuffleWith(elements, n => this.nextIntBetween(0, n));
  }
}
_a = TRandomTypeId;
/** @internal */
export const live = /*#__PURE__*/Layer.effect(Tag, /*#__PURE__*/core.commit( /*#__PURE__*/core.map(seed => new TRandomImpl(seed))( /*#__PURE__*/tRef.make( /*#__PURE__*/new Random.PCGRandom(Math.random() * 4294967296 >>> 0).getState()))));
/** @internal */
export const next = /*#__PURE__*/core.flatMap(Tag, random => random.next);
/** @internal */
export const nextBoolean = /*#__PURE__*/core.flatMap(Tag, random => random.nextBoolean);
/** @internal */
export const nextInt = /*#__PURE__*/core.flatMap(Tag, random => random.nextInt);
/** @internal */
export const nextIntBetween = (low, high) => core.flatMap(Tag, random => random.nextIntBetween(low, high));
/** @internal */
export const nextRange = (min, max) => core.flatMap(Tag, random => random.nextRange(min, max));
/** @internal */
export const shuffle = elements => core.flatMap(Tag, random => random.shuffle(elements));
