// @effect/stm/THub.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.unbounded = exports.subscribeScoped = exports.subscribe = exports.sliding = exports.size = exports.publishAll = exports.publish = exports.isShutdown = exports.isFull = exports.isEmpty = exports.dropping = exports.capacity = exports.bounded = exports.awaitShutdown = exports.THubTypeId = void 0;
var internal = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("@effect/stm/internal/tHub"));
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
/**
 * @since 1.0.0
 * @category symbols
 */
const THubTypeId = internal.THubTypeId;
/**
 * Waits until the hub is shutdown. The `STM` returned by this method will
 * not resume until the queue has been shutdown. If the hub is already
 * shutdown, the `STM` will resume right away.
 *
 * @since 1.0.0
 * @category mutations
 */
exports.THubTypeId = THubTypeId;
const awaitShutdown = internal.awaitShutdown;
/**
 * Creates a bounded hub with the back pressure strategy. The hub will retain
 * messages until they have been taken by all subscribers, applying back
 * pressure to publishers if the hub is at capacity.
 *
 * @since 1.0.0
 * @category constructors
 */
exports.awaitShutdown = awaitShutdown;
const bounded = internal.bounded;
/**
 * Returns the number of elements the hub can hold.
 *
 * @since 1.0.0
 * @category getters
 */
exports.bounded = bounded;
const capacity = internal.capacity;
/**
 * Creates a bounded hub with the dropping strategy. The hub will drop new
 * messages if the hub is at capacity.
 *
 * @since 1.0.0
 * @category constructors
 */
exports.capacity = capacity;
const dropping = internal.dropping;
/**
 * Returns `true` if the `THub` contains zero elements, `false` otherwise.
 *
 * @since 1.0.0
 * @category getters
 */
exports.dropping = dropping;
const isEmpty = internal.isEmpty;
/**
 * Returns `true` if the `THub` contains at least one element, `false`
 * otherwise.
 *
 * @since 1.0.0
 * @category getters
 */
exports.isEmpty = isEmpty;
const isFull = internal.isFull;
/**
 * Returns `true` if `shutdown` has been called, otherwise returns `false`.
 *
 * @since 1.0.0
 * @category getters
 */
exports.isFull = isFull;
const isShutdown = internal.isShutdown;
/**
 * Publishes a message to the hub, returning whether the message was published
 * to the hub.
 *
 * @since 1.0.0
 * @category mutations
 */
exports.isShutdown = isShutdown;
const publish = internal.publish;
/**
 * Publishes all of the specified messages to the hub, returning whether they
 * were published to the hub.
 *
 * @since 1.0.0
 * @category mutations
 */
exports.publish = publish;
const publishAll = internal.publishAll;
/**
 * Retrieves the size of the hub, which is equal to the number of elements
 * in the hub. This may be negative if fibers are suspended waiting for
 * elements to be added to the hub.
 *
 * @since 1.0.0
 * @category getters
 */
exports.publishAll = publishAll;
const size = internal.size;
/**
 * Creates a bounded hub with the sliding strategy. The hub will add new
 * messages and drop old messages if the hub is at capacity.
 *
 * For best performance use capacities that are powers of two.
 *
 * @since 1.0.0
 * @category constructors
 */
exports.size = size;
const sliding = internal.sliding;
/**
 * Subscribes to receive messages from the hub. The resulting subscription can
 * be evaluated multiple times to take a message from the hub each time. The
 * caller is responsible for unsubscribing from the hub by shutting down the
 * queue.
 *
 * @since 1.0.0
 * @category mutations
 */
exports.sliding = sliding;
const subscribe = internal.subscribe;
/**
 * Subscribes to receive messages from the hub. The resulting subscription can
 * be evaluated multiple times within the scope to take a message from the hub
 * each time.
 *
 * @since 1.0.0
 * @category mutations
 */
exports.subscribe = subscribe;
const subscribeScoped = internal.subscribeScoped;
/**
 * Creates an unbounded hub.
 *
 * @since 1.0.0
 * @category constructors
 */
exports.subscribeScoped = subscribeScoped;
const unbounded = internal.unbounded;
exports.unbounded = unbounded;
