// @effect/stm/TSet.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.union = exports.transformSTM = exports.transform = exports.toReadonlySet = exports.toReadonlyArray = exports.toHashSet = exports.toChunk = exports.takeSomeSTM = exports.takeSome = exports.takeFirstSTM = exports.takeFirst = exports.size = exports.retainIfDiscard = exports.retainIf = exports.removeIfDiscard = exports.removeIf = exports.removeAll = exports.remove = exports.reduceSTM = exports.reduce = exports.make = exports.isEmpty = exports.intersection = exports.has = exports.fromIterable = exports.forEach = exports.empty = exports.difference = exports.add = exports.TSetTypeId = void 0;
var internal = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("@effect/stm/internal/tSet"));
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
/**
 * @since 1.0.0
 * @category symbols
 */
const TSetTypeId = internal.TSetTypeId;
/**
 * Stores new element in the set.
 *
 * @since 1.0.0
 * @category mutations
 */
exports.TSetTypeId = TSetTypeId;
const add = internal.add;
/**
 * Atomically transforms the set into the difference of itself and the
 * provided set.
 *
 * @since 1.0.0
 * @category mutations
 */
exports.add = add;
const difference = internal.difference;
/**
 * Makes an empty `TSet`.
 *
 * @since 1.0.0
 * @category constructors
 */
exports.difference = difference;
const empty = internal.empty;
/**
 * Atomically performs transactional-effect for each element in set.
 *
 * @since 1.0.0
 * @category elements
 */
exports.empty = empty;
const forEach = internal.forEach;
/**
 * Makes a new `TSet` initialized with provided iterable.
 *
 * @since 1.0.0
 * @category constructors
 */
exports.forEach = forEach;
const fromIterable = internal.fromIterable;
/**
 * Tests whether or not set contains an element.
 *
 * @since 1.0.0
 * @category elements
 */
exports.fromIterable = fromIterable;
const has = internal.has;
/**
 * Atomically transforms the set into the intersection of itself and the
 * provided set.
 *
 * @since 1.0.0
 * @category mutations
 */
exports.has = has;
const intersection = internal.intersection;
/**
 * Tests if the set is empty or not
 *
 * @since 1.0.0
 * @category getters
 */
exports.intersection = intersection;
const isEmpty = internal.isEmpty;
/**
 * Makes a new `TSet` that is initialized with specified values.
 *
 * @since 1.0.0
 * @category constructors
 */
exports.isEmpty = isEmpty;
const make = internal.make;
/**
 * Atomically folds using a pure function.
 *
 * @since 1.0.0
 * @category folding
 */
exports.make = make;
const reduce = internal.reduce;
/**
 * Atomically folds using a transactional function.
 *
 * @since 1.0.0
 * @category folding
 */
exports.reduce = reduce;
const reduceSTM = internal.reduceSTM;
/**
 * Removes a single element from the set.
 *
 * @since 1.0.0
 * @category mutations
 */
exports.reduceSTM = reduceSTM;
const remove = internal.remove;
/**
 * Removes elements from the set.
 *
 * @since 1.0.0
 * @category mutations
 */
exports.remove = remove;
const removeAll = internal.removeAll;
/**
 * Removes bindings matching predicate and returns the removed entries.
 *
 * @since 1.0.0
 * @category mutations
 */
exports.removeAll = removeAll;
const removeIf = internal.removeIf;
/**
 * Removes elements matching predicate.
 *
 * @since 1.0.0
 * @category mutations
 */
exports.removeIf = removeIf;
const removeIfDiscard = internal.removeIfDiscard;
/**
 * Retains bindings matching predicate and returns removed bindings.
 *
 * @since 1.0.0
 * @category mutations
 */
exports.removeIfDiscard = removeIfDiscard;
const retainIf = internal.retainIf;
/**
 * Retains elements matching predicate.
 *
 * @since 1.0.0
 * @category mutations
 */
exports.retainIf = retainIf;
const retainIfDiscard = internal.retainIfDiscard;
/**
 * Returns the set's cardinality.
 *
 * @since 1.0.0
 * @category getters
 */
exports.retainIfDiscard = retainIfDiscard;
const size = internal.size;
/**
 * Takes the first matching value, or retries until there is one.
 *
 * @since 1.0.0
 * @category mutations
 */
exports.size = size;
const takeFirst = internal.takeFirst;
/**
 * Takes the first matching value, or retries until there is one.
 *
 * @since 1.0.0
 * @category mutations
 */
exports.takeFirst = takeFirst;
const takeFirstSTM = internal.takeFirstSTM;
/**
 * Takes all matching values, or retries until there is at least one.
 *
 * @since 1.0.0
 * @category mutations
 */
exports.takeFirstSTM = takeFirstSTM;
const takeSome = internal.takeSome;
/**
 * Takes all matching values, or retries until there is at least one.
 *
 * @since 1.0.0
 * @category mutations
 */
exports.takeSome = takeSome;
const takeSomeSTM = internal.takeSomeSTM;
/**
 * Collects all elements into a `Chunk`.
 *
 * @since 1.0.0
 * @category destructors
 */
exports.takeSomeSTM = takeSomeSTM;
const toChunk = internal.toChunk;
/**
 * Collects all elements into a `HashSet`.
 *
 * @since 1.0.0
 * @category destructors
 */
exports.toChunk = toChunk;
const toHashSet = internal.toHashSet;
/**
 * Collects all elements into a `ReadonlyArray`.
 *
 * @since 1.0.0
 * @category destructors
 */
exports.toHashSet = toHashSet;
const toReadonlyArray = internal.toReadonlyArray;
/**
 * Collects all elements into a `ReadonlySet`.
 *
 * @since 1.0.0
 * @category destructors
 */
exports.toReadonlyArray = toReadonlyArray;
const toReadonlySet = internal.toReadonlySet;
/**
 * Atomically updates all elements using a pure function.
 *
 * @since 1.0.0
 * @category mutations
 */
exports.toReadonlySet = toReadonlySet;
const transform = internal.transform;
/**
 * Atomically updates all elements using a transactional function.
 *
 * @since 1.0.0
 * @category mutations
 */
exports.transform = transform;
const transformSTM = internal.transformSTM;
/**
 * Atomically transforms the set into the union of itself and the provided
 * set.
 *
 * @since 1.0.0
 * @category mutations
 */
exports.transformSTM = transformSTM;
const union = internal.union;
exports.union = union;
