// @effect/stm/mjs/THub.mjs
import * as internal from "@effect/stm/internal/tHub";
/**
 * @since 1.0.0
 * @category symbols
 */
export const THubTypeId = internal.THubTypeId;
/**
 * Waits until the hub is shutdown. The `STM` returned by this method will
 * not resume until the queue has been shutdown. If the hub is already
 * shutdown, the `STM` will resume right away.
 *
 * @since 1.0.0
 * @category mutations
 */
export const awaitShutdown = internal.awaitShutdown;
/**
 * Creates a bounded hub with the back pressure strategy. The hub will retain
 * messages until they have been taken by all subscribers, applying back
 * pressure to publishers if the hub is at capacity.
 *
 * @since 1.0.0
 * @category constructors
 */
export const bounded = internal.bounded;
/**
 * Returns the number of elements the hub can hold.
 *
 * @since 1.0.0
 * @category getters
 */
export const capacity = internal.capacity;
/**
 * Creates a bounded hub with the dropping strategy. The hub will drop new
 * messages if the hub is at capacity.
 *
 * @since 1.0.0
 * @category constructors
 */
export const dropping = internal.dropping;
/**
 * Returns `true` if the `THub` contains zero elements, `false` otherwise.
 *
 * @since 1.0.0
 * @category getters
 */
export const isEmpty = internal.isEmpty;
/**
 * Returns `true` if the `THub` contains at least one element, `false`
 * otherwise.
 *
 * @since 1.0.0
 * @category getters
 */
export const isFull = internal.isFull;
/**
 * Returns `true` if `shutdown` has been called, otherwise returns `false`.
 *
 * @since 1.0.0
 * @category getters
 */
export const isShutdown = internal.isShutdown;
/**
 * Publishes a message to the hub, returning whether the message was published
 * to the hub.
 *
 * @since 1.0.0
 * @category mutations
 */
export const publish = internal.publish;
/**
 * Publishes all of the specified messages to the hub, returning whether they
 * were published to the hub.
 *
 * @since 1.0.0
 * @category mutations
 */
export const publishAll = internal.publishAll;
/**
 * Retrieves the size of the hub, which is equal to the number of elements
 * in the hub. This may be negative if fibers are suspended waiting for
 * elements to be added to the hub.
 *
 * @since 1.0.0
 * @category getters
 */
export const size = internal.size;
/**
 * Creates a bounded hub with the sliding strategy. The hub will add new
 * messages and drop old messages if the hub is at capacity.
 *
 * For best performance use capacities that are powers of two.
 *
 * @since 1.0.0
 * @category constructors
 */
export const sliding = internal.sliding;
/**
 * Subscribes to receive messages from the hub. The resulting subscription can
 * be evaluated multiple times to take a message from the hub each time. The
 * caller is responsible for unsubscribing from the hub by shutting down the
 * queue.
 *
 * @since 1.0.0
 * @category mutations
 */
export const subscribe = internal.subscribe;
/**
 * Subscribes to receive messages from the hub. The resulting subscription can
 * be evaluated multiple times within the scope to take a message from the hub
 * each time.
 *
 * @since 1.0.0
 * @category mutations
 */
export const subscribeScoped = internal.subscribeScoped;
/**
 * Creates an unbounded hub.
 *
 * @since 1.0.0
 * @category constructors
 */
export const unbounded = internal.unbounded;
