// @effect/stm/mjs/TSemaphore.mjs
/**
 * @since 1.0.0
 */
import * as internal from "@effect/stm/internal/tSemaphore";
/**
 * @since 1.0.0
 * @category symbols
 */
export const TSemaphoreTypeId = internal.TSemaphoreTypeId;
/**
 * @since 1.0.0
 * @category mutations
 */
export const acquire = internal.acquire;
/**
 * @since 1.0.0
 * @category mutations
 */
export const acquireN = internal.acquireN;
/**
 * @since 1.0.0
 * @category getters
 */
export const available = internal.available;
/**
 * @since 1.0.0
 * @category constructors
 */
export const make = internal.make;
/**
 * @since 1.0.0
 * @category mutations
 */
export const release = internal.release;
/**
 * @since 1.0.0
 * @category mutations
 */
export const releaseN = internal.releaseN;
/**
 * @since 1.0.0
 * @category mutations
 */
export const withPermit = internal.withPermit;
/**
 * @since 1.0.0
 * @category mutations
 */
export const withPermits = internal.withPermits;
/**
 * @since 1.0.0
 * @category mutations
 */
export const withPermitScoped = internal.withPermitScoped;
/**
 * @since 1.0.0
 * @category mutations
 */
export const withPermitsScoped = internal.withPermitsScoped;
/**
 * @since 1.0.0
 * @category unsafe
 */
export const unsafeMake = internal.unsafeMakeSemaphore;
