// @effect/stm/internal/opCodes/stm.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.OP_WITH_STM_RUNTIME = exports.OP_TRACED = exports.OP_SYNC = exports.OP_SUCCEED = exports.OP_RETRY = exports.OP_PROVIDE = exports.OP_ON_SUCCESS = exports.OP_ON_RETRY = exports.OP_ON_FAILURE = exports.OP_INTERRUPT = exports.OP_FAIL = exports.OP_DIE = void 0;
/** @internal */
const OP_WITH_STM_RUNTIME = "WithSTMRuntime";
/** @internal */
exports.OP_WITH_STM_RUNTIME = OP_WITH_STM_RUNTIME;
const OP_ON_FAILURE = "OnFailure";
/** @internal */
exports.OP_ON_FAILURE = OP_ON_FAILURE;
const OP_ON_RETRY = "OnRetry";
/** @internal */
exports.OP_ON_RETRY = OP_ON_RETRY;
const OP_ON_SUCCESS = "OnSuccess";
/** @internal */
exports.OP_ON_SUCCESS = OP_ON_SUCCESS;
const OP_PROVIDE = "Provide";
/** @internal */
exports.OP_PROVIDE = OP_PROVIDE;
const OP_SYNC = "Sync";
/** @internal */
exports.OP_SYNC = OP_SYNC;
const OP_SUCCEED = "Succeed";
/** @internal */
exports.OP_SUCCEED = OP_SUCCEED;
const OP_RETRY = "Retry";
/** @internal */
exports.OP_RETRY = OP_RETRY;
const OP_FAIL = "Fail";
/** @internal */
exports.OP_FAIL = OP_FAIL;
const OP_DIE = "Die";
/** @internal */
exports.OP_DIE = OP_DIE;
const OP_INTERRUPT = "Interrupt";
/** @internal */
exports.OP_INTERRUPT = OP_INTERRUPT;
const OP_TRACED = "Traced";
exports.OP_TRACED = OP_TRACED;
